File: backend/app/services/project_service.py
"""Project service."""

from typing import Optional
from uuid import UUID

from app.models.project import Project, ProjectStatus
from app.repositories.client_repository import ClientRepository
from app.repositories.professional_repository import ProfessionalRepository
from app.repositories.project_repository import ProjectRepository
from app.schemas.project import ProjectCreate, ProjectUpdate


class ProjectService:
    """Project service class."""

    def __init__(
        self,
        project_repository: ProjectRepository,
        client_repository: ClientRepository,
        professional_repository: ProfessionalRepository,
    ):
        """Initialize service."""
        self.project_repository = project_repository
        self.client_repository = client_repository
        self.professional_repository = professional_repository

    async def create_project(
        self, client_id: UUID, project_data: ProjectCreate
    ) -> Project:
        """Create a new project."""
        # Verify client exists
        client = self.client_repository.get_by_id(client_id)
        if not client:
            raise ValueError("Client profile not found")

        return self.project_repository.create(client_id, project_data)

    async def get_project_by_id(self, project_id: UUID) -> Optional[Project]:
        """Get project by ID."""
        return self.project_repository.get_by_id(project_id)

    async def update_project(
        self, project_id: UUID, project_data: ProjectUpdate
    ) -> Optional[Project]:
        """Update project."""
        return self.project_repository.update(project_id, project_data)

    async def delete_project(self, project_id: UUID) -> bool:
        """Delete project."""
        return self.project_repository.delete(project_id)

    async def list_projects(self, skip: int = 0, limit: int = 20) -> list[Project]:
        """List projects."""
        return self.project_repository.list_projects(skip=skip, limit=limit)

    async def count_projects(self) -> int:
        """Count total projects."""
        return self.project_repository.count_projects()

    async def get_projects_by_client(
        self, client_id: UUID, skip: int = 0, limit: int = 20
    ) -> list[Project]:
        """Get projects by client ID."""
        return self.project_repository.get_projects_by_client(
            client_id, skip=skip, limit=limit
        )

    async def get_projects_by_status(
        self, status: ProjectStatus, skip: int = 0, limit: int = 20
    ) -> list[Project]:
        """Get projects by status."""
        return self.project_repository.get_projects_by_status(
            status, skip=skip, limit=limit
        )

    async def get_projects_by_professional(
        self, professional_id: UUID, skip: int = 0, limit: int = 20
    ) -> list[Project]:
        """Get projects by professional ID."""
        return self.project_repository.get_projects_by_professional(
            professional_id, skip=skip, limit=limit
        )

    async def get_open_projects(self, skip: int = 0, limit: int = 20) -> list[Project]:
        """Get all open projects."""
        return self.project_repository.get_open_projects(skip=skip, limit=limit)

    async def assign_professional(
        self, project_id: UUID, professional_id: UUID
    ) -> Optional[Project]:
        """Assign a professional to a project."""
        # Verify project exists
        project = self.project_repository.get_by_id(project_id)
        if not project:
            raise ValueError("Project not found")

        # Verify project is open
        if project.status != ProjectStatus.OPEN:
            raise ValueError(
                f"Project is not open for assignment. Current status: {project.status}"
            )

        # Verify professional exists
        professional = self.professional_repository.get_by_id(professional_id)
        if not professional:
            raise ValueError("Professional profile not found")

        return self.project_repository.assign_professional(project_id, professional_id)

    async def update_status(
        self, project_id: UUID, status: ProjectStatus
    ) -> Optional[Project]:
        """Update project status."""
        # Verify project exists
        project = self.project_repository.get_by_id(project_id)
        if not project:
            raise ValueError("Project not found")

        # Validate status transitions
        if (
            project.status == ProjectStatus.COMPLETED
            and status != ProjectStatus.COMPLETED
        ):
            raise ValueError("Cannot change status of completed project")

        if (
            project.status == ProjectStatus.CANCELLED
            and status != ProjectStatus.CANCELLED
        ):
            raise ValueError("Cannot change status of cancelled project")

        return self.project_repository.update_status(project_id, status)

    async def complete_project(self, project_id: UUID) -> Optional[Project]:
        """Mark project as completed."""
        return await self.update_status(project_id, ProjectStatus.COMPLETED)

    async def cancel_project(self, project_id: UUID) -> Optional[Project]:
        """Mark project as cancelled."""
        return await self.update_status(project_id, ProjectStatus.CANCELLED)

File: frontend/src/services/projectService.ts
import apiClient from './api'
import { Project, ProjectCreate, ProjectUpdate, ProjectFilters } from '@/types/project'

export interface PaginatedProjects {
  items: Project[]
  total: number
  page: number
  size: number
  pages: number
}

export const projectService = {
  async getProjects(filters?: ProjectFilters): Promise<PaginatedProjects> {
    const params = new URLSearchParams()

    // Backend expects 'status_filter' not 'status'
    if (filters?.status) params.append('status_filter', filters.status)
    if (filters?.client_id) params.append('client_id', filters.client_id)
    if (filters?.professional_id) params.append('professional_id', filters.professional_id)
    if (filters?.only_open !== undefined) params.append('only_open', String(filters.only_open))
    if (filters?.skip !== undefined) params.append('skip', String(filters.skip))
    if (filters?.limit !== undefined) params.append('limit', String(filters.limit))

    const response = await apiClient.get<PaginatedProjects>(`/projects?${params.toString()}`)
    return response.data
  },

  async getProjectById(id: string): Promise<Project> {
    const response = await apiClient.get<Project>(`/projects/${id}`)
    return response.data
  },

  async getProjectsByProfessional(professionalId: string, status?: string): Promise<Project[]> {
    const filters: ProjectFilters = {
      professional_id: professionalId,
      limit: 100,
    }

    const response = await this.getProjects(filters)

    // Filter by status if provided
    if (status) {
      return response.items.filter(p => p.status === status)
    }

    return response.items
  },

  async createProject(data: ProjectCreate): Promise<Project> {
    const response = await apiClient.post<Project>('/projects', data)
    return response.data
  },

  async updateProject(id: string, data: ProjectUpdate): Promise<Project> {
    const response = await apiClient.put<Project>(`/projects/${id}`, data)
    return response.data
  },

  async deleteProject(id: string): Promise<void> {
    await apiClient.delete(`/projects/${id}`)
  },

  async completeProject(id: string): Promise<Project> {
    const response = await apiClient.patch<Project>(`/projects/${id}/complete`)
    return response.data
  },

  async cancelProject(id: string): Promise<Project> {
    const response = await apiClient.patch<Project>(`/projects/${id}/cancel`)
    return response.data
  },
}
